fairscale/experimental/wgit/signal_sparsity_profiling.py

```diff
@@ -0,0 +1,62 @@
+# Copyright (c) Facebook, Inc. and its affiliates. All rights reserved.
+#
+# This source code is licensed under the BSD license found in the
+# LICENSE file in the root directory of this source tree.
+
+from typing import List
+
+import torch
+from torch import Tensor
+
+
+class EnergyConcentrationProfile:
+    """Compute "energy" concentration level for a tensor
+
+    Args:
+        dim (int):
+            The dimension to measure.
+        top_k_percents (List[float]):
+            List of percentage values. For each value, the `measure`
+            function will compute and return the percentage of "energy"
+            concentrated on that top-K percent of values in the dimension
+            to measure. Note, this is the opposite of the sparsity percentage.
+    """
+
+    def __init__(self, dim: int, top_k_percents: List[float]) -> None:
+        assert isinstance(dim, int)
+        self.dim = dim
+        self.percents = []
+        last_p = 0.0
+        for p in top_k_percents:
+            assert isinstance(p, (int, float))
+            assert p > 0, p
+            assert p <= 100, p
+            assert p > last_p, f"p {p} should be larger than last_p {last_p}"
+            self.percents.append(float(p))
+            last_p = p
+
+    def measure(self, in_tensor: Tensor) -> List[Tensor]:
+        """Compute the return the results
+
+            Note, we want this function to be nonblocking and async.
```

Comment:
I was wondering, how are we making this function non-blocking and async? Is it simply by the fact we are not using `torch.cuda.synchronize()` anywhere within the function? Also, what was the goal for making this specifically async?

Reply:
good questions. anything that need computation results will causing blocking, like print() or tensor.items().

The goal for making this async is that we may do this profiling in a loop for many tensors. If this is sync, then things will be really slow.

thanks for the review!